### PR TITLE
move LowerCaseString to common for common use

### PR DESCRIPTION
**Pull request by wbpcode**
Signed-off-by: wbpcode <comems@msn.com>

This PR moved `Envoy::Http::LowerCaseString` to common so that other modules can implement case insensitive string comparison without deps on Http. 

This PR is second sub PR of #16049.

Commit Message:
Additional Description:
Risk Level: Low
Testing: N/A
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

**Comment by wrowe** [comment]
It seems clang_tidy simply wants a cleaner constructor here;
```
Writing fixes to clang-tidy-fixes.yaml ...
clang-tidy check failed, potentially fixed by clang-apply-replacements:
Diagnostics:
- BuildDirectory: /build/tmp/_bazel_envoybuild/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy
  DiagnosticMessage:
    FileOffset: 675
    FilePath: include/envoy/common/case_insensitive.h
    Message: use '= default' to define a trivial copy constructor
    Replacements:
    - {FilePath: include/envoy/common/case_insensitive.h, Length: 21, Offset: 723,
      ReplacementText: ''}
    - {FilePath: include/envoy/common/case_insensitive.h, Length: 2, Offset: 744,
      ReplacementText: = default;}
  DiagnosticName: modernize-use-equals-default
  Level: Warning
- BuildDirectory: /build/tmp/_bazel_envoybuild/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy
  DiagnosticMessage:
    FileOffset: 767
    FilePath: include/envoy/common/case_insensitive.h
    Message: use '= default' to define a trivial copy-assignment operator
    Replacements:
    - {FilePath: include/envoy/common/case_insensitive.h, Length: 50, Offset: 806,
      ReplacementText: = default;}
  DiagnosticName: modernize-use-equals-default
  Level: Warning
```

File: test/extensions/filters/http/fault/fault_filter_integration_test.cc [hunk 1]
```diff
@@ -3,6 +3,8 @@
 
 #include "gtest/gtest.h"
 
+using Envoy::Http::HeaderValueOf;
```
Comment (lizan) [inline test/extensions/filters/http/fault/fault_filter_integration_test.cc:6]: ?

Comment (wbpcode) [reply test/extensions/filters/http/fault/fault_filter_integration_test.cc:6]: This `using` is used to ensure tests of fault filter can be passed. 
I haven't found the reason why the original fault filter test passed. It originally neither used `using` nor `Http::HeaderValueOf`, but directly used `HeaderValueOf`. In my opinion, because the namespace is completely different, it should not be able to use this name directly.

Comment (lizan) [reply test/extensions/filters/http/fault/fault_filter_integration_test.cc:6]: I know but you didn't change the test content, so why you needed to add this here?

Comment (wbpcode) [reply test/extensions/filters/http/fault/fault_filter_integration_test.cc:6]: Because when I reworked the `LowerCaseString`, the test of the fault filter made an error during the compilation, and it prompted me to change the `HeaderValueOf` to `Http::HeaderValueOf`.

This prompt is correct, but I can't explain why the compilation can pass before modifying `LowerCaseString`.

Comment (wbpcode) [reply test/extensions/filters/http/fault/fault_filter_integration_test.cc:6]: @lizan I get it.

Before modifying `Http::LowerCaseString`, `Http::LowerCaseString` is normal class in `Http` namespace. The `HeaderValueOf` call in fault filter test can be handled correctly because ADL. Check https://en.cppreference.com/w/cpp/language/adl.

After modifying `Http::LowerCaseString`, `Http::LowerCaseString` become a specialization of `Envoy::ValidatedLowerCaseStr` template which changed the lookup scope of `HeaderValueOf`. So we must use `using` declaration or udpate `HeaderValueOf` to `Http::HeaderValueOf`

**Comment by wbpcode** [comment]
@lizan I think this PR is ready for review.

File: include/envoy/common/case_insensitive.h [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+protected:
+  void lower() {
+    std::transform(string_.begin(), string_.end(), string_.begin(), absl::ascii_tolower);
+  }
+
+  std::string string_;
+};
+
+template <bool (*V)(absl::string_view)> class ValidatedLowerCaseStr : public LowerCaseStrBase {
```
Comment (lizan) [inline include/envoy/common/case_insensitive.h:45]: I'm not sure why we need this to be a template.

Comment (wbpcode) [reply include/envoy/common/case_insensitive.h:45]: @lizan Because I want to get a protocol-independent lower case string and be able to set a different validator. The template is used so that the validator is pluggable.

Here are some discussions about this question in original PR #16049. 
@jmarantz  https://github.com/envoyproxy/envoy/pull/16049#discussion_r623008023

Comment (lizan) [reply include/envoy/common/case_insensitive.h:45]: OK I see, then this should be commented why it is template. and in Http, it should be named like validateHttpLowerCaseString.

Comment (jmarantz) [reply include/envoy/common/case_insensitive.h:45]: TBH my recollection from the.other PR is that there are a few jumbled concepts here and I want to just quickly mention my line of thinking.

IMO if you really need a validator,  forget whether there's a compelling reason to prefer
 * template
 * virtual method
 * function-pointer
But if we do use a function-pointer then in IMO it could be a std::function rather than a templatized function pointer.

Also I think once we are plugging in validation, there's no compelling reason to use LowerCaseString just for Tracing's sake, unless there's something about Tracing that wants strings to be lower-cased. IDK whether that's the case.

That said, I have no fundamental objection to moving LowerCaseString to common as long as it doesn't make anything slower for HTTP headers, which are performance critical.

I don't understand why LowerCaseString has to be paired with a validator though. Can't you just leave LowerCaseString exactly like it is, and move it out of http into common? That seems like a legitimate concept on its own. You could make another layer that adds validation if this is helpful in some way.

Comment (wbpcode) [reply include/envoy/common/case_insensitive.h:45]: @lizan I will add some new comments to explain why we need this template.
In Http, I kept the original Http::LowerCaseString name so as not to affect the existing code as much as possible. After all, the name Http::LowerCaseString is already widely used.

Comment (jmarantz) [reply include/envoy/common/case_insensitive.h:45]: So this isn't a move...it's a copy?

There's nothing inherently wrong with doing a move and changing all the references though that will make the PR large, so you could always do the move of the code and leave a forwarding reference in Http, then follow up with a cleanup.

But I don't understand why the Trace functionality in the other PR needs lower-casing at all.

Comment (wbpcode) [reply include/envoy/common/case_insensitive.h:45]: @jmarantz 
- First, about the performance: This modification will not affect the performance of Http::LowerCaseString since there is no new logic. The introduction of templates may increase the compilation overhead slightly.
- Second, about the relationship between validator and LowerCaseString: I provide a class `LowerCaseStrBase` as a common implementation of LowerCaseString which with no any validation and can be used directly. Then I provide `ValidatedLowerCaseStr` template just as a helper for who wants to add validation to LowerCaseString, for example, Http. 
So that means I've got two layers as you think. One layer is the plain `LowerCaseString`, and the second layer is the `ValidatedLowerCaseStr` that can mount a different validator.

Comment (wbpcode) [reply include/envoy/common/case_insensitive.h:45]: @jmarantz At the source code level, it is move.
I mean that in Http I kept the previous name `LowerCaseString` instead of renaming it to `ValidatedHttpLowerCaseString` or something like that to make PR clean.

Comment (jmarantz) [reply include/envoy/common/case_insensitive.h:50]: OK...sorry, TBH I'm skeptical about the overall strategy and also concerned there may be overhead added to http headers.

But for the moment maybe it'd be better to focus on strategy. Do you want to describe the high level of what you are trying to achieve in a bug or a google doc? It might be easier to iterate there.

Comment (wbpcode) [reply include/envoy/common/case_insensitive.h:50]: Thanks for you comments and patience. @jmarantz 

In the original PR #16049, there is the high level design of the general tracing. And I remember I had a more detailed discussion with you in slack earlier.

For now, I think we can try to focus on this PR first. Because this PR don't involve the design of the new general tracing yet, just some pre-work. This is also the reason why I split the original PR into multiple PRs.
I think we can discuss the design of general tracing in detail in the new PR (One or two weeks after the completion of this PR). 

For this PR I think the most important points are the following:
* whether the PR is clean or not.
* whether the PR is able to keep the existing functionality working.
* whether the PR can guarantee that the existing performance is not affected.
* Whether the PR achieves its purpose: to strip LowerCaseString from Http and get a general LowerCaseString.

**Comment by wbpcode** [comment]
@lizan Some comments are added. 

I personally think it might be better to keep the name `Http::LowerCaseString` in Http as it is already widely used. Keeping the original name by `using` would make this PR much cleaner.

File: source/extensions/tracers/common/ot/opentracing_driver_impl.cc [hunk 3]
```diff
@@ -29,7 +29,7 @@ class OpenTracingHTTPHeadersWriter : public opentracing::HTTPHeadersWriter {
   // opentracing::HTTPHeadersWriter
   opentracing::expected<void> Set(opentracing::string_view key,
                                   opentracing::string_view value) const override {
-    Http::LowerCaseString lowercase_key{key};
+    Http::LowerCaseString lowercase_key{{key.data(), key.size()}};
```
Comment (jmarantz) [inline source/extensions/tracers/common/ot/opentracing_driver_impl.cc:32]: should this be moved to a different PR? Doesn't seem like this should be needed in this one.

**Comment by wbpcode** [comment]
Here is a more detailed doc for general tracing. https://docs.google.com/document/d/1gxk5vfHl-Gt3wiNZ-f46Dwmk2NFiEzFJQugmZmpSXk8/edit?usp=sharing
@jmarantz Thank you very much for your continued patience and help. 🌷 

Personally, I still think this PR is actually not that closely related to general tracing.
Because all people who want to use the lower case semantics, but do not want to rely on Http can benefit from it. It is not specifically designed for general tracing.

**Comment by wbpcode** [comment]
Personally, I still think this PR is actually not that closely related to general tracing.
Because all people who want to use the lower case semantics, but do not want to rely on Http can benefit from it. It is not specifically designed for general tracing.

**Comment by jmarantz** [comment]
@wbpcode agreed. This PR seems like a good one to apply if we need such a semantic in the future. There's nothing wrong with moving it, and this PR would be a good start to do that. But I'm not sure if there'd be any more users at this point, so maybe we can put this one on hold till then?

And I still think your C-style function-pointer based validation could (a) be a separate PR from the move and (b) could be a lambda for easier use in some cases.
